specs/keccak256/variable-length-input.md

```diff
@@ -0,0 +1,122 @@
+
+If the last part of the inputs is less than 136 bytes, a padding must be applied. The padding is a multi-rate padding, it pads single bit 1 followed by minimum number of bits 0 followed by a single bit 1, such that the result is 136 bytes.
+
+## Circuit behavior
+
+### Lookup Region
+
+This region splits the input to multiple parts, each part corresponds to their `Keccak-f` permutation round.
+This is also a lookup table for the other circuits to lookup the Keccak256 input to the output.
+
+Define:
+
+We use `Round` for the cells of the following columns in a row, it represents the information in a `keccak_f` round.
+
+The `curr: Round` and `next: Round` represent the current and the next row.
+
+Columns:
+
+- `hash_id` Sequential number for grouping different hash
+- `input_len` Length for correct padding
+- `input` 136 bytes to be absorbed in this round
+- `acc_len` How many length we have absorbed
+- `acc_input` Accumulatd bytes by random linear combination (in big-endian order)
+- `output` The base-2 `state[:4]` output from this round `keccak_f`
+- `is_end_result` This flag indicates the hash output is final. The purpose is for external circuit to lookup.
+
+Selector:
+
+- `is_last_round_of_circuit` Last round of the whole circuit (not hash), to avoid `next` wrap around.
+
+#### Checks
+
+1. `hash_id` is sequential `next.id - curr.id in [0, 1]`
+2. If we are not in the last round of the circuit or last round of permutation
+   1. Checks the `input_len` is  `curr.input_len === next.input_len`
+   2. `next.acc_len === curr.acc_len + 136`
+   3. `next.acc_len <= curr.input_len`
+   4. `next.acc_input === curr.acc_input * (r**136) + RLC(input, r)`
+3. If we are in the last round of the circuit or in last round of permutation
+   1. Checks the accumulation ends here `assert (curr.input_len - curr.acc_len) in range(136)`
+   2. Clear the variables:  `next.acc_len === 0`, `next.acc_input === 0`
+4. `is_end_result === next.id - curr.id`
+
+#### Lookup
+
+To lookup the Keccak256 input to the output, query the following columns:
+
+- `is_end_result`: bool
+- `input`: RLC
+- `output`: RLC
+
+and constrain `is_end_result === 1`
+
+### Padding Region
+
+Note that we define a new `acc_len` which increament byte by byte, where the `acc_len` in the lookup region bumps by 136 bytes.
+
+#### Plain behavior: The padding rule
+
+```python
+def get_padding(acc_len) -> bytes:
+    """
+    output big-endian bytes
+    """
+    # note that padlen is at maximum 135
+    padlen = 136 - acc_len 
+    if padlen == 0:
+        # no pad
+        return []
+    elif padlen == 1:
+        # pad 0b10000001
+        return [0x81]
+    elif 1 < padlen < 136:
+        return [0x80] + ([0x00] * int(padlen - 2)) + [0x01]
+    else:
+        raise ValueError("unreachable")
+```
+
+#### Circuit
+
+The Padding Region is a 136-row region.
```

Comment:
I started implementing like 17 rows of 8bytes + 1 word for each. So 17 rows of 9 columns (with a few extra columns for the accomulators). Makes it much easier for the constraint appliance.

Can we copy into another region the required data to make the lookups? Or will increase the costs too much?

Reply:
Ohhh. 

I started coding like:
- 17 rows of 9 columns
- Copy each byte to the padding_config region where we have the following columns:
- `byte`
- `total_len`
- `acc_len`
- `pad_len`
- `running_sum`

And is in this region where we will copy all bytes from offset 0-136 or even 0-MAX_INP_LEN as they have to be sequential for the entire hash.

WDYT?

Reply:
The constraints are the same. The only thing that changes is how we layout the data. I did it like this as simplifies a lot more the things form the impl side. 

But maybe this is not inline with your proposal, and then I'm happy to refactor

Reply:
Make sense to me. As long as the padding_config is 1 col 136 rows it should be fine. It's okay to copy `byte` col in padding_config to a 17-rows-9 columns gadget for other processes.

Reply:
Can't only be this. Needs some padding extra fields as I said.
Otherwise, you miss info to make the constraints between bytes.

Config looks like:
```rust
/// Constrains each byte of the keccak input to be padded correctly.
#[derive(Debug, Clone)]
pub(crate) struct PaddingBuilderConfig<F> {
    q_enable: Selector,
    // Contains:
    // - `byte` at Rotation(0)
    // - `pad_byte` at Rotation(1)
    byte: Column<Advice>,
    total_len: Column<Advice>,
    acc_len: Column<Advice>,
    // Contains:
    // - `pad_len` at Rotation(0)
    // - `pad_len_inv` at Rotation(1)
    pad_len: Column<Advice>,
    running_sum: Column<Advice>,
}
```
I can change this and avoid using rotations, having more columns and less rows.
Feel free to decide on this! :)

Reply:
Can we do "constraints between bytes" by copying them to 17 word-gadgets?

Reply:
I'm not sure I follow you here. 
The idea i was following (what I understood in our last call) was something along the lines of:

- Witness all of the bytes and padd correctly in `PermutationConfig`.
- From the previous config, we copy the info required to check that a byte is correctly padded into `PaddingBuilderConfig`.
- Finally, once all these constraints are applied, we copy each 8 bytes to the WordBuilder together with the word they form to constraint word construction correctness as well as that the bytes are indeed between [0-255].

Then we return the RLC accomulated for each byte of the permutation + the resulting State words.

Reply:
Resolved offline. We agreed to go naive and defer the optimization later.
The optimization is about merging `byte` and `pad_byte` and using `Rotation(2)` to reduce columns and with the cost of increasing rotation, which is more cost-efficient in the aggregation circuit. But merging the columns would make debugging gates harder and optimization at this stage is a bit early.